[Windows] path to kotlinc not quoted

Until now I only used kscript in unix and within stuff like WSL, but as I'm currently reevaluating kscript for additional use, I've noticed I can now install it on Windows and use it directly. Sadly, out of the box it will break:

The PC I was installing it on has a pretty usual setup: 
- kscript installed via scoop today
- using IntelliJ's builtin Kotlin distribution (per default installed to "C:\\Program Files\JetBrains\IntelliJ<..>\plugins\Kotlin\kotlinc\bin\kotlinc.bat" (set in PATH in my case)

In Git Bash the log output to the README's hello world example is as follows (cmd meanwhile flies completely off the handle):

```
$ ./hello_kscript.kts
[kscript] [ERROR] Compilation of scriplet failed:
[kscript] [ERROR] Command     : 'cmd /c C:[bs]Program Files[bs]JetBrains[bs]IntelliJ IDEA Community Edition 2021.1.1[bs]plugins[bs]Kotlin[bs]kotlinc[bs]bin[bs]kotlinc.bat   -d "C:[bs]Users[bs]chris[bs]AppData[bs]Local[bs]kscript[bs]jar_79d1195f5b541eb5aa1a07f6bec7e715[bs]scriplet.jar" "C:[bs]Users[bs]chris[bs]AppData[bs]Local[bs]kscript[bs]jar_79d1195f5b541eb5aa1a07f6bec7e715[bs]Hello_kscript.kts" "C:[bs]Users[bs]chris[bs]AppData[bs]Local[bs]kscript[bs]jar_79d1195f5b541eb5aa1a07f6bec7e715[bs]Main_Hello_kscript.kt"'
[kscript] [ERROR] Exit Code   : 1
[kscript] [ERROR] Stdout      : ''
[kscript] [ERROR] Stderr      : 'Der Befehl "C:[bs]Program" ist entweder falsch geschrieben oder[nl]konnte nicht gefunden werden.[nl]'
[kscript] [ERROR]
```

Error message is in german but it comes down to the 'Program Files' directory containing a space in Windows, which breaks the path into two segments.


A quick lookup suggests CommandResolver::resolveKotlinBinary being the culprit - this argument of the command is the only one not being quoted. Escaping would be an alternative (but that would seem inconsistent behavior to me).

Would a PR for this be welcome?



Update/
It seems to be even worse. Directly quoting the path is out, the ShellExecutor.kt in kscript should be probably using something else than `cmd /c "<path>" <args>`, because the first character being a quote will cause legacy behavior:
https://stackoverflow.com/a/356014/1907778

There are multiple ways to deal with this problem, one being to just not use absolute paths, but that's a major refactoring with way too much potential regressions.

The seemingly hackiest but somewhat most isolated and simplest solution seems to be to both put quotes around <path> BUT ALSO prefix it with a single space. It all seems to work from there, and we can isolate this behavior inside `CommandResolver` only, and only for Windows. A space will disable the unexpected dequoting behavior, and a quoted path can contain as many spaces as you'd want.
I will open a PR to show.

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -308,7 +308,7 @@ dependencies {
     implementation("net.igsoft:tablevis:0.6.0")
     implementation("io.arrow-kt:arrow-core:1.1.2")
 
-    implementation("io.github.kscripting:shell:0.5.0")
+    implementation("io.github.kscripting:shell:0.5.1")
 
     implementation("org.slf4j:slf4j-nop:2.0.5")
 

--- a/src/main/kotlin/io/github/kscripting/kscript/resolver/CommandResolver.kt
+++ b/src/main/kotlin/io/github/kscripting/kscript/resolver/CommandResolver.kt
@@ -137,12 +137,7 @@ class CommandResolver(val osConfig: OsConfig) {
             osConfig.kotlinHomeDir.resolve("bin", if (osConfig.osType.isWindowsLike()) "$binary.bat" else binary)
                 .convert(osConfig.osType)
                 .stringPath()
-        return if (osConfig.osType == OsType.WINDOWS) {
-            // if the first character in args in `cmd /c <args>` is a quote, cmd will remove it as well as the
-            // last quote character within args before processing the term, which removes our quotes.
-            " ${filePathQuotationMark}${pathToKotlinc}${filePathQuotationMark}"
-        } else {
-            pathToKotlinc
-        }
+        val quotes = if (osConfig.osType == OsType.WINDOWS) filePathQuotationMark else ""
+        return "${quotes}${pathToKotlinc}${quotes}"
     }
 }

--- a/src/main/kotlin/io/github/kscripting/kscript/resolver/CommandResolver.kt
+++ b/src/main/kotlin/io/github/kscripting/kscript/resolver/CommandResolver.kt
@@ -12,6 +12,12 @@ class CommandResolver(val osConfig: OsConfig) {
     private val classPathSeparator =
         if (osConfig.osType.isWindowsLike() || osConfig.osType.isPosixHostedOnWindows()) ";" else ":"
 
+    private val filePathQuotationMark = when (osConfig.osType) {
+        OsType.WINDOWS -> '"'
+        else -> '\''
+    }
+
+
     fun getKotlinJreVersion(): String {
         val kotlin = resolveKotlinBinary("kotlin")
         return "$kotlin -version"
@@ -90,41 +96,23 @@ class CommandResolver(val osConfig: OsConfig) {
     private fun resolveKotlinOpts(kotlinOpts: Set<KotlinOpt>) = kotlinOpts.joinToString(" ") { it.value }
 
     private fun resolveCompilerOpts(compilerOpts: Set<CompilerOpt>) = compilerOpts.joinToString(" ") { it.value }
-
     private fun resolveJarFile(jar: OsPath): String {
-        val jarFileQuotationMark: Char = when (osConfig.osType) {
-            OsType.WINDOWS -> '"'
-            else -> '\''
-        }
-
-        return "${jarFileQuotationMark}${resolveQuotedPath(jar)}${jarFileQuotationMark}"
+        return "${filePathQuotationMark}${resolveQuotedPath(jar)}${filePathQuotationMark}"
     }
 
     private fun resolveFiles(filePaths: Set<OsPath>): String {
-        val filePathQuotationMark: Char = when (osConfig.osType) {
-            OsType.WINDOWS -> '"'
-            else -> '\''
-        }
-
         return filePaths.joinToString(" ") {
-            "${filePathQuotationMark}${
-                resolveQuotedPath(it)
-            }${filePathQuotationMark}"
+            "${filePathQuotationMark}${resolveQuotedPath(it)}${filePathQuotationMark}"
         }
     }
 
     private fun resolveUserArgs(userArgs: List<String>): String {
-        val userArgQuotationMark: Char = when (osConfig.osType) {
-            OsType.WINDOWS -> '"'
-            else -> '\''
-        }
-
         return userArgs.joinToString(" ") {
-            "${userArgQuotationMark}${
+            "${filePathQuotationMark}${
                 it.replace(
                     "\"", "\\\""
                 )
-            }${userArgQuotationMark}"
+            }${filePathQuotationMark}"
         }
     }
 
@@ -133,14 +121,11 @@ class CommandResolver(val osConfig: OsConfig) {
             return ""
         }
 
-        val classpathParameterQuotationMark: Char = when (osConfig.osType) {
-            OsType.WINDOWS -> '"'
-            else -> '\''
-        }
-
-        val classpath = classpathParameterQuotationMark + dependencies.joinToString(classPathSeparator) {
-            resolveQuotedPath(it)
-        } + classpathParameterQuotationMark
+        val classpath = "${filePathQuotationMark}${
+            dependencies.joinToString(classPathSeparator) {
+                resolveQuotedPath(it)
+            }
+        }${filePathQuotationMark}"
 
         return "-classpath $classpath"
     }
@@ -148,8 +133,16 @@ class CommandResolver(val osConfig: OsConfig) {
     private fun resolveQuotedPath(osPath: OsPath): String = osPath.toNativeOsPath().stringPath()
 
     private fun resolveKotlinBinary(binary: String): String {
-        return osConfig.kotlinHomeDir.resolve("bin", if (osConfig.osType.isWindowsLike()) "$binary.bat" else binary)
-            .convert(osConfig.osType)
-            .stringPath()
+        val pathToKotlinc =
+            osConfig.kotlinHomeDir.resolve("bin", if (osConfig.osType.isWindowsLike()) "$binary.bat" else binary)
+                .convert(osConfig.osType)
+                .stringPath()
+        return if (osConfig.osType == OsType.WINDOWS) {
+            // if the first character in args in `cmd /c <args>` is a quote, cmd will remove it as well as the
+            // last quote character within args before processing the term, which removes our quotes.
+            " ${filePathQuotationMark}${pathToKotlinc}${filePathQuotationMark}"
+        } else {
+            pathToKotlinc
+        }
     }
 }

--- a/src/test/kotlin/io/github/kscripting/kscript/resolver/CommandResolverTest.kt
+++ b/src/test/kotlin/io/github/kscripting/kscript/resolver/CommandResolverTest.kt
@@ -23,11 +23,11 @@ class CommandResolverTest {
         val commandResolver = CommandResolver(config.osConfig)
 
         assertThat(commandResolver.compileKotlin(jarPath, depPaths, filePaths, compilerOpts)).isEqualTo(
-            """ "C:\Users\Admin With Spaces\scoop\apps\kotlin\current\bin\kotlinc.bat" -abc -def --experimental -classpath "C:\My Workspace\Code\.m2\somepath\dep1.jar;C:\My Workspace\Code\.m2\somepath\dep2.jar;C:\My Workspace\Code\.m2\somepath\dep3.jar" -d "C:\My Workspace\Code\.kscript\cache\somefile.jar" "C:\My Workspace\Code\source\somepath\dep1.kt" "C:\My Workspace\Code\source\somepath\dep2.kts""""
+            """"C:\Users\Admin With Spaces\scoop\apps\kotlin\current\bin\kotlinc.bat" -abc -def --experimental -classpath "C:\My Workspace\Code\.m2\somepath\dep1.jar;C:\My Workspace\Code\.m2\somepath\dep2.jar;C:\My Workspace\Code\.m2\somepath\dep3.jar" -d "C:\My Workspace\Code\.kscript\cache\somefile.jar" "C:\My Workspace\Code\source\somepath\dep1.kt" "C:\My Workspace\Code\source\somepath\dep2.kts""""
         )
 
         assertThat(commandResolver.executeKotlin(jarArtifact, depPaths, userArgs, kotlinOpts)).isEqualTo(
-            """ "C:\Users\Admin With Spaces\scoop\apps\kotlin\current\bin\kotlin.bat" -k1 -k2 --disable -classpath "C:\My Workspace\Code\.m2\somepath\dep1.jar;C:\My Workspace\Code\.m2\somepath\dep2.jar;C:\My Workspace\Code\.m2\somepath\dep3.jar;C:\My Workspace\Code\.kscript\cache\somefile.jar;C:\Users\Admin With Spaces\scoop\apps\kotlin\current\lib\kotlin-script-runtime.jar" mainClass "arg" "u" "ments""""
+            """"C:\Users\Admin With Spaces\scoop\apps\kotlin\current\bin\kotlin.bat" -k1 -k2 --disable -classpath "C:\My Workspace\Code\.m2\somepath\dep1.jar;C:\My Workspace\Code\.m2\somepath\dep2.jar;C:\My Workspace\Code\.m2\somepath\dep3.jar;C:\My Workspace\Code\.kscript\cache\somefile.jar;C:\Users\Admin With Spaces\scoop\apps\kotlin\current\lib\kotlin-script-runtime.jar" mainClass "arg" "u" "ments""""
         )
     }
 

--- a/src/test/kotlin/io/github/kscripting/kscript/resolver/CommandResolverTest.kt
+++ b/src/test/kotlin/io/github/kscripting/kscript/resolver/CommandResolverTest.kt
@@ -18,16 +18,16 @@ class CommandResolverTest {
         val (config, jarPath, jarArtifact, depPaths, filePaths) = createTestData(
             OsType.WINDOWS,
             "C:\\My Workspace\\Code",
-            "C:\\Users\\Admin\\scoop\\apps\\kotlin\\current"
+            "C:\\Users\\Admin With Spaces\\scoop\\apps\\kotlin\\current"
         )
         val commandResolver = CommandResolver(config.osConfig)
 
         assertThat(commandResolver.compileKotlin(jarPath, depPaths, filePaths, compilerOpts)).isEqualTo(
-            """C:\Users\Admin\scoop\apps\kotlin\current\bin\kotlinc.bat -abc -def --experimental -classpath "C:\My Workspace\Code\.m2\somepath\dep1.jar;C:\My Workspace\Code\.m2\somepath\dep2.jar;C:\My Workspace\Code\.m2\somepath\dep3.jar" -d "C:\My Workspace\Code\.kscript\cache\somefile.jar" "C:\My Workspace\Code\source\somepath\dep1.kt" "C:\My Workspace\Code\source\somepath\dep2.kts""""
+            """ "C:\Users\Admin With Spaces\scoop\apps\kotlin\current\bin\kotlinc.bat" -abc -def --experimental -classpath "C:\My Workspace\Code\.m2\somepath\dep1.jar;C:\My Workspace\Code\.m2\somepath\dep2.jar;C:\My Workspace\Code\.m2\somepath\dep3.jar" -d "C:\My Workspace\Code\.kscript\cache\somefile.jar" "C:\My Workspace\Code\source\somepath\dep1.kt" "C:\My Workspace\Code\source\somepath\dep2.kts""""
         )
 
         assertThat(commandResolver.executeKotlin(jarArtifact, depPaths, userArgs, kotlinOpts)).isEqualTo(
-            """C:\Users\Admin\scoop\apps\kotlin\current\bin\kotlin.bat -k1 -k2 --disable -classpath "C:\My Workspace\Code\.m2\somepath\dep1.jar;C:\My Workspace\Code\.m2\somepath\dep2.jar;C:\My Workspace\Code\.m2\somepath\dep3.jar;C:\My Workspace\Code\.kscript\cache\somefile.jar;C:\Users\Admin\scoop\apps\kotlin\current\lib\kotlin-script-runtime.jar" mainClass "arg" "u" "ments""""
+            """ "C:\Users\Admin With Spaces\scoop\apps\kotlin\current\bin\kotlin.bat" -k1 -k2 --disable -classpath "C:\My Workspace\Code\.m2\somepath\dep1.jar;C:\My Workspace\Code\.m2\somepath\dep2.jar;C:\My Workspace\Code\.m2\somepath\dep3.jar;C:\My Workspace\Code\.kscript\cache\somefile.jar;C:\Users\Admin With Spaces\scoop\apps\kotlin\current\lib\kotlin-script-runtime.jar" mainClass "arg" "u" "ments""""
         )
     }
 

